File: apps/android/app/src/test/java/app/murmur/android/SttCoverageTest.kt
package app.murmur.android

import app.murmur.android.audio.Wav
import app.murmur.android.stt.TimedSpan
import app.murmur.android.stt.TranscribeOutput
import app.murmur.android.stt.adaptiveThreshold
import app.murmur.android.stt.assessCoverage
import app.murmur.android.stt.findCutPoint
import app.murmur.android.stt.lastVoicedSec
import app.murmur.android.stt.mergeTranscripts
import app.murmur.android.stt.spansFromVerbose
import app.murmur.android.stt.transcribeComplete
import kotlinx.coroutines.runBlocking
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/** Mirrors apps/desktop/tests/stt-coverage.test.ts. */
class SttCoverageTest {
    private val rate = 16_000
    private val threshold = -48.0

    /** Deterministic noise bursts (speech) and near-silence (pauses) at the given times. */
    private fun clip(totalSec: Double, speech: List<Pair<Double, Double>>): ShortArray {
        val pcm = ShortArray((totalSec * rate).toInt())
        var seed = 12345L
        fun rnd(): Double {
            seed = (seed * 1103515245L + 12345L) and 0x7fffffffL
            return seed / 0x7fffffff.toDouble() - 0.5
        }
        for (i in pcm.indices) pcm[i] = (rnd() * 20).toInt().toShort()
        for ((from, to) in speech) {
            for (i in (from * rate).toInt() until minOf(pcm.size, (to * rate).toInt())) pcm[i] = (rnd() * 16000).toInt().toShort()
        }
        return pcm
    }

    private fun out(text: String, lastEnd: Double? = null) =
        TranscribeOutput(text = text, latencyMs = 10, spans = lastEnd?.let { listOf(TimedSpan(0.0, it)) })

    private fun wavSeconds(wav: ByteArray): Double = Wav.decodePcm16(wav).first.size / rate.toDouble()

    private val head = "This is an Android app. Please implement this in a clean manner and Wispr Flow."
    private val tail = "It should also keep the pill visible while the keyboard is up."
    private val pcm = clip(20.0, listOf(0.3 to 8.0, 8.6 to 19.7))
    private val speechEnd = lastVoicedSec(pcm, rate, threshold)

    @Test
    fun `coverage from timings, word rate and empty answers`() {
        assertFalse(assessCoverage(out("hello there how are you", 19.6), 20.0).truncated)
        val cut = assessCoverage(out("please implement this and Wispr Flow", 7.9), 60.0)
        assertTrue(cut.truncated)
        assertEquals("timestamps", cut.reason)
        assertEquals(52.1, cut.missingSec, 0.05)
        assertFalse(assessCoverage(out("a few words here", 8.2), 10.0).truncated)
        val sparse = assessCoverage(out("only eleven words came back for a very long recording here"), 60.0)
        assertTrue(sparse.truncated)
        assertEquals("rate", sparse.reason)
        assertNull(sparse.transcribedSec)
        assertFalse(assessCoverage(out((1..40).joinToString(" ") { "word" }), 20.0).truncated)
        assertFalse(assessCoverage(out("send it"), 5.0).truncated)
        val empty = assessCoverage(out(""), 12.0)
        assertTrue(empty.truncated)
        assertEquals("empty", empty.reason)
        assertEquals(0.0, empty.transcribedSec)
        assertFalse(assessCoverage(out(""), 1.5).truncated)
    }

    @Test
    fun `voice activity helpers find the end of speech and a threshold above the noise floor`() {
        assertEquals(19.7, speechEnd, 0.05)
        assertTrue(adaptiveThreshold(pcm, rate, -48.0) >= -48.0)
        assertEquals(-1.0, lastVoicedSec(ShortArray(rate), rate, threshold), 0.0)
    }

    @Test
    fun `cut point sits in the pause after the point where the transcript stopped`() {
        val cut = findCutPoint(pcm, rate, 7.9, threshold)
        assertTrue("cut $cut", cut > 8.05 && cut < 8.55)
        assertEquals(13.9, findCutPoint(pcm, rate, 14.0, threshold), 0.01)
    }

    @Test
    fun `merge drops the words the tail repeats from the seam but not one short shared word`() {
        assertEquals(
            "Please implement this in a clean manner and Wispr Flow. It should also use the predictive back gesture.",
            mergeTranscripts(
                "Please implement this in a clean manner and Wispr Flow.",
                "Wispr Flow. It should also use the predictive back gesture."
            )
        )
        assertEquals("send it to the the report today", mergeTranscripts("send it to the", "the report today"))
        assertEquals("we talked about kubernetes is hard", mergeTranscripts("we talked about kubernetes", "kubernetes is hard"))
        assertEquals("tail", mergeTranscripts("", "tail"))
        assertEquals("head", mergeTranscripts("head", "  "))
    }

    @Test
    fun `spans prefer word timings over segment timings`() {
        val json = JSONObject(
            """{"text":"x","segments":[{"start":0,"end":30,"no_speech_prob":0.1}],"words":[{"word":"x","start":0.5,"end":7.9}]}"""
        )
        assertEquals(listOf(TimedSpan(0.5, 7.9)), spansFromVerbose(json))
        assertEquals(listOf(TimedSpan(0.0, 30.0)), spansFromVerbose(JSONObject("""{"segments":[{"start":0,"end":30}]}""")))
        assertNull(spansFromVerbose(JSONObject("""{"segments":[{"no_speech_prob":0.1}]}""")))
    }

    @Test
    fun `resumes from the pause after the point where the prompted transcript stopped`() = runBlocking {
        val calls = ArrayList<Pair<Double, String?>>()
        val r = transcribeComplete(
            pcm, rate, speechEnd, threshold,
            prompt = "Vocabulary: Wispr Flow. Dictation with punctuation.",
            tailPrompt = "Dictation with punctuation."
        ) { wav, prompt ->
            val seconds = wavSeconds(wav)
            calls.add(seconds to prompt)
            // The vocabulary prompt makes the model stop right after "Wispr Flow".
            if (prompt?.contains("Vocabulary") == true) out(head, 7.9) else out(tail, seconds - 0.2)
        }
        assertEquals("$head $tail", r.output.text)
        assertEquals(1, r.resumed)
        assertTrue(r.recoveredSec > 10)
        assertFalse(r.coverage.truncated)
        assertEquals(2, calls.size)
        assertEquals("Dictation with punctuation.", calls[1].second)
        // The tail starts inside the 8.0-8.6 s pause, so no word is cut in half.
        val cut = 20 - calls[1].first
        assertTrue("cut $cut", cut > 8.0 && cut < 8.6)
    }

    @Test
    fun `makes no extra request when the transcript covers the speech`() = runBlocking {
        var calls = 0
        val r = transcribeComplete(pcm, rate, speechEnd, threshold, prompt = "p", tailPrompt = null) { _, _ ->
            calls++
            out(head, 19.6)
        }
        assertEquals(1, calls)
        assertEquals(0, r.resumed)
        assertEquals(head, r.output.text)
    }

    @Test
    fun `retries the whole clip without the prompt when there are no timings to resume from`() = runBlocking {
        val prompts = ArrayList<String?>()
        val r = transcribeComplete(pcm, rate, speechEnd, threshold, prompt = "Vocabulary: x. Dictation.", tailPrompt = null) { _, prompt ->
            prompts.add(prompt)
            if (prompt != null) out("only a few words came back") else out("$head $tail")
        }
        assertEquals(listOf("Vocabulary: x. Dictation.", null), prompts)
        assertEquals("$head $tail", r.output.text)
        assertEquals(1, r.resumed)
        // ...but a retry that is not clearly fuller loses to the first answer.
        val same = transcribeComplete(pcm, rate, speechEnd, threshold, prompt = "p", tailPrompt = null) { _, prompt ->
            if (prompt != null) out("one two three four five") else out("one two three four six")
        }
        assertEquals("one two three four five", same.output.text)
        assertEquals(0, same.resumed)
    }

    @Test
    fun `never loses the head when the resume request fails`() = runBlocking {
        var calls = 0
        val r = transcribeComplete(pcm, rate, speechEnd, threshold, prompt = "p", tailPrompt = null) { _, _ ->
            calls++
            if (calls == 1) out(head, 7.9) else throw IllegalStateException("rate limited")
        }
        assertEquals(head, r.output.text)
        assertEquals(0, r.resumed)
        assertTrue(r.coverage.truncated)
    }

    @Test
    fun `chains several resumes when the recognizer keeps stopping early`() = runBlocking {
        val pieces = listOf("first part.", "second part.", "third part.")
        var calls = 0
        val r = transcribeComplete(pcm, rate, speechEnd, threshold, prompt = null, tailPrompt = null, maxRounds = 3) { wav, _ ->
            val seconds = wavSeconds(wav)
            val i = calls++
            // Each answer covers only the first half of whatever audio it was given.
            out(pieces.getOrElse(i) { "more." }, if (i == 0) 6.5 else minOf(seconds - 0.2, seconds / 2))
        }
        assertTrue("calls $calls", calls in 3..4)
        assertTrue(r.output.text.startsWith("first part. second part. third part."))
        assertTrue(r.resumed >= 2)
    }
}

File: apps/desktop/tests/stt-coverage.test.ts
import { describe, expect, it } from 'vitest'
import { analyze } from '@core/audio/vad'
import { decodeWavPcm16 } from '@core/audio/wav'
import {
  assessCoverage,
  findCutPoint,
  mergeTranscripts,
  transcribeComplete,
  type Transcribe
} from '@core/stt/coverage'
import type { TranscribeOutput } from '@core/stt/types'

const RATE = 16000
const THRESHOLD = -48

/** Deterministic noise bursts (speech) and near-silence (pauses) at the given times. */
function clip(totalSec: number, speech: Array<[number, number]>): Int16Array {
  const pcm = new Int16Array(Math.round(totalSec * RATE))
  let seed = 12345
  const rnd = (): number => {
    seed = (seed * 1103515245 + 12345) & 0x7fffffff
    return seed / 0x7fffffff - 0.5
  }
  for (let i = 0; i < pcm.length; i++) pcm[i] = Math.round(rnd() * 20)
  for (const [from, to] of speech) {
    for (let i = Math.round(from * RATE); i < Math.min(pcm.length, Math.round(to * RATE)); i++)
      pcm[i] = Math.round(rnd() * 16000)
  }
  return pcm
}

const out = (
  text: string,
  lastEnd?: number,
  extra: Partial<TranscribeOutput> = {}
): TranscribeOutput => ({
  text,
  latencyMs: 10,
  spans: lastEnd === undefined ? undefined : [{ start: 0, end: lastEnd }],
  ...extra
})

const wavSeconds = (wav: Uint8Array): number => decodeWavPcm16(wav).pcm.length / RATE

describe('assessCoverage', () => {
  it('accepts a transcript whose timings reach the end of the speech', () => {
    const c = assessCoverage(out('hello there how are you', 19.6), 20)
    expect(c.truncated).toBe(false)
    expect(c.transcribedSec).toBe(19.6)
  })
  it('flags a transcript that stops well before the speech ends', () => {
    const c = assessCoverage(out('please implement this in a clean manner and Wispr Flow', 7.9), 60)
    expect(c.truncated).toBe(true)
    expect(c.reason).toBe('timestamps')
    expect(c.missingSec).toBeCloseTo(52.1, 1)
  })
  it('tolerates timing jitter and short trailing pauses', () => {
    expect(assessCoverage(out('a few words here', 8.2), 10).truncated).toBe(false)
    expect(assessCoverage(out('a few words here', 27.5), 30).truncated).toBe(false)
  })
  it('prefers the last word end over a bogus segment end', () => {
    const spans = [
      { start: 0, end: 30 },
      { start: 0.5, end: 7.9 }
    ]
    // Whichever span ends last counts; a segment running to its window end still means 30 s covered.
    expect(assessCoverage({ text: 'some words', spans }, 60).transcribedSec).toBe(30)
  })
  it('falls back to the word rate when there are no timings', () => {
    const sparse = assessCoverage(
      out('only eleven words came back for a very long recording here'),
      60
    )
    expect(sparse.truncated).toBe(true)
    expect(sparse.reason).toBe('rate')
    expect(sparse.transcribedSec).toBeNull()
    const dense = assessCoverage(out(Array(40).fill('word').join(' ')), 20)
    expect(dense.truncated).toBe(false)
    // Short clips are never judged by rate: two words in five seconds is a normal dictation.
    expect(assessCoverage(out('send it'), 5).truncated).toBe(false)
  })
  it('treats an empty transcript for real speech as everything missing', () => {
    const c = assessCoverage(out(''), 12)
    expect(c.truncated).toBe(true)
    expect(c.reason).toBe('empty')
    expect(c.transcribedSec).toBe(0)
    expect(assessCoverage(out(''), 1.5).truncated).toBe(false)
  })
})

describe('findCutPoint', () => {
  const pcm = clip(20, [
    [0.3, 8.0],
    [8.6, 19.7]
  ])
  it('cuts in the middle of the pause after the point where the transcript stopped', () => {
    const cut = findCutPoint(pcm, RATE, 7.9, { thresholdDb: THRESHOLD })
    expect(cut).toBeGreaterThan(8.05)
    expect(cut).toBeLessThan(8.55)
  })
  it('cuts just before the target when the speaker never paused', () => {
    const cut = findCutPoint(pcm, RATE, 14, { thresholdDb: THRESHOLD })
    expect(cut).toBeCloseTo(13.9, 2)
  })
})

describe('mergeTranscripts', () => {
  it('drops the words the tail repeats from the seam', () => {
    expect(
      mergeTranscripts(
        'Please implement this in a clean manner and Wispr Flow.',
        'Wispr Flow. It should also use the predictive back gesture.'
      )
    ).toBe(
      'Please implement this in a clean manner and Wispr Flow. It should also use the predictive back gesture.'
    )
  })
  it('does not treat one short shared word as overlap', () => {
    expect(mergeTranscripts('send it to the', 'the report today')).toBe(
      'send it to the the report today'
    )
    expect(mergeTranscripts('we talked about kubernetes', 'kubernetes is hard')).toBe(
      'we talked about kubernetes is hard'
    )
  })
  it('handles empty sides', () => {
    expect(mergeTranscripts('', 'tail')).toBe('tail')
    expect(mergeTranscripts('head', '  ')).toBe('head')
  })
})

describe('transcribeComplete', () => {
  const HEAD =
    'This is an Android app. It should take advantage of the predictive back gesture. Please implement this in a clean manner and Wispr Flow.'
  const TAIL = 'It should also keep the pill visible while the keyboard is up.'
  const pcm = clip(20, [
    [0.3, 8.0],
    [8.6, 19.7]
  ])
  const speechEnd = analyze(pcm, { sampleRate: RATE, thresholdDb: THRESHOLD }).lastVoicedMs / 1000

  it('resumes from the pause after the point where the prompted transcript stopped', async () => {
    const calls: Array<{ seconds: number; prompt?: string }> = []
    const transcribe: Transcribe = async (wav, prompt) => {
      const seconds = wavSeconds(wav)
      calls.push({ seconds, prompt })
      // The vocabulary prompt makes the model stop right after "Wispr Flow".
      if (prompt?.includes('Vocabulary')) return out(HEAD, 7.9)
      return out(TAIL, seconds - 0.2)
    }
    const r = await transcribeComplete(transcribe, {
      pcm,
      sampleRate: RATE,
      speechEndSec: speechEnd,
      thresholdDb: THRESHOLD,
      prompt: 'Vocabulary: Wispr Flow. Dictation with punctuation.',
      tailPrompt: 'Dictation with punctuation.'
    })
    expect(r.text).toBe(`${HEAD} ${TAIL}`)
    expect(r.resumed).toBe(1)
    expect(r.recoveredSec).toBeGreaterThan(10)
    expect(r.coverage.truncated).toBe(false)
    expect(calls).toHaveLength(2)
    expect(calls[1].prompt).toBe('Dictation with punctuation.')
    // The tail starts inside the 8.0-8.6 s pause, so no word is cut in half.
    const cut = 20 - calls[1].seconds
    expect(cut).toBeGreaterThan(8.0)
    expect(cut).toBeLessThan(8.6)
    expect(r.spans?.[r.spans.length - 1].end).toBeGreaterThan(19)
  })

  it('makes no extra request when the transcript covers the speech', async () => {
    let calls = 0
    const r = await transcribeComplete(
      async () => {
        calls++
        return out(HEAD, 19.6)
      },
      { pcm, sampleRate: RATE, speechEndSec: speechEnd, thresholdDb: THRESHOLD, prompt: 'p' }
    )
    expect(calls).toBe(1)
    expect(r.resumed).toBe(0)
    expect(r.text).toBe(HEAD)
  })

  it('retries the whole clip without the prompt when there are no timings to resume from', async () => {
    const calls: Array<string | undefined> = []
    const r = await transcribeComplete(
      async (_wav, prompt) => {
        calls.push(prompt)
        return prompt ? out('only a few words came back') : out(`${HEAD} ${TAIL}`)
      },
      {
        pcm,
        sampleRate: RATE,
        speechEndSec: speechEnd,
        thresholdDb: THRESHOLD,
        prompt: 'Vocabulary: Wispr Flow. Dictation with punctuation.'
      }
    )
    expect(calls).toEqual(['Vocabulary: Wispr Flow. Dictation with punctuation.', undefined])
    expect(r.text).toBe(`${HEAD} ${TAIL}`)
    expect(r.resumed).toBe(1)
  })

  it('keeps the shorter answer when the retry is not clearly fuller', async () => {
    const r = await transcribeComplete(
      async (_wav, prompt) =>
        prompt ? out('one two three four five') : out('one two three four six'),
      { pcm, sampleRate: RATE, speechEndSec: speechEnd, thresholdDb: THRESHOLD, prompt: 'p' }
    )
    expect(r.text).toBe('one two three four five')
    expect(r.resumed).toBe(0)
  })

  it('recovers speech the model returned nothing for', async () => {
    const r = await transcribeComplete(
      async (wav, prompt) => (prompt ? out('', 0) : out(TAIL, wavSeconds(wav) - 0.3)),
      { pcm, sampleRate: RATE, speechEndSec: speechEnd, thresholdDb: THRESHOLD, prompt: 'p' }
    )
    expect(r.text).toBe(TAIL)
    expect(r.resumed).toBe(1)
  })

  it('never loses the head when the resume request fails', async () => {
    let calls = 0
    const r = await transcribeComplete(
      async () => {
        calls++
        if (calls === 1) return out(HEAD, 7.9)
        throw new Error('rate limited')
      },
      { pcm, sampleRate: RATE, speechEndSec: speechEnd, thresholdDb: THRESHOLD, prompt: 'p' }
    )
    expect(r.text).toBe(HEAD)
    expect(r.resumed).toBe(0)
    expect(r.coverage.truncated).toBe(true)
  })

  it('stops when a resumed tail brings no timings and no further progress', async () => {
    let calls = 0
    const r = await transcribeComplete(
      async () => {
        calls++
        return calls === 1 ? out(HEAD, 7.9) : out(TAIL)
      },
      { pcm, sampleRate: RATE, speechEndSec: speechEnd, thresholdDb: THRESHOLD, prompt: 'p' }
    )
    expect(calls).toBe(2)
    expect(r.text).toBe(`${HEAD} ${TAIL}`)
    expect(r.resumed).toBe(1)
  })

  it('chains several resumes when the recognizer keeps stopping early', async () => {
    const pieces = ['first part.', 'second part.', 'third part.']
    let calls = 0
    const r = await transcribeComplete(
      async (wav) => {
        const seconds = wavSeconds(wav)
        const i = calls++
        // Each answer covers only the first third of whatever audio it was given.
        return out(pieces[i] ?? 'more.', i === 0 ? 6.5 : Math.min(seconds - 0.2, seconds / 2))
      },
      { pcm, sampleRate: RATE, speechEndSec: speechEnd, thresholdDb: THRESHOLD, maxRounds: 3 }
    )
    expect(calls).toBeGreaterThanOrEqual(3)
    expect(calls).toBeLessThanOrEqual(4)
    expect(r.text.startsWith('first part. second part. third part.')).toBe(true)
    expect(r.resumed).toBeGreaterThanOrEqual(2)
  })
})
